kube-runtime/src/manager.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{
+    controller::{controller, trigger_owners, trigger_self, trigger_with, Context, Error, ReconcilerAction},
+    reflector::{
+        reflector,
+        store::{Store, Writer},
+        ObjectRef,
+    },
+    utils::{try_flatten_applied, try_flatten_touched},
+    watcher::{self, watcher},
+};
+use futures::{stream, Stream, TryFuture};
+use kube::api::{Api, ListParams, Meta};
+use serde::de::DeserializeOwned;
+use std::pin::Pin;
+
+/// A builder for controller
+pub struct ControllerBuilder<K>
+where
+    K: Clone + Meta + 'static,
+{
+    // NB: Need to Unpin for stream::select_all
+    // TODO: get an arbitrary std::error::Error in here?
+    selector: Vec<Pin<Box<dyn Stream<Item = Result<ObjectRef<K>, watcher::Error>>>>>,
+    reader: Store<K>,
+}
+
+impl<K> ControllerBuilder<K>
+where
+    K: Clone + Meta + DeserializeOwned + 'static,
+{
+    /// Create a ControllerBuilder on a type `K`
+    ///
+    /// Configure `ListParams` and `Api` so you only get reconcile events
+    /// for the correct Api scope (cluster/all/namespaced), or ListParams subset
+    ///
+    /// A writer is exposed for convenience so you can peak into the main reflector's state.
+    pub fn new(owned_api: Api<K>, lp: ListParams, writer: Writer<K>) -> Self {
```

Comment:
It's unsafe to have multiple writers, so I'm not sure we gain much by being able to supply our own store. Maybe just have a method for retrieving the `Store` instead? Then again, that wouldn't be very clean for reconcilers that wanted to access the store.. :/

```rust
let builder = ControllerBuilder::new(cms, ListParams::default());
let store = builder.store();
builder.run(|obj| /* do stuff */, err_policy, ctx)
```

Reply:
i'm all for encapsulating it. If people need to write more stuff into it I'm leaning towards making better entrypoints for external injection points like `watches`.

Reply:
Mhm, at that point you're deep enough into the woods anyway that you'd probably be better served by "hard-mode" `controller()` anyway.